go/store/nbs/reflog_ring_buffer.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nbs
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// errUnsafeIteration is returned when iterating through a ring buffer too slowly and new, inserted data is detected
+// as wrapping around into the iteration range.
+var errUnsafeIteration = errors.New(
+	"unable to finish iteration: insertion index has wrapped around into iteration range")
+
+// reflogRootHashEntry is a data container for a root hash update that was recorded to the chunk journal. It contains
+// the root and the time at which it was written.
+type reflogRootHashEntry struct {
+	root      string
+	timestamp time.Time
+}
+
+// reflogRingBuffer is a fixed size circular buffer that allows the most recent N entries to be iterated over (where
+// N is equal to the size requested when this ring buffer is constructed. Its locking strategy assumes that
+// only new entries are written to the head (through Push) and that existing entries will never need to be
+// updated. Internally, it allocates a slice that is twice as large as the requested size, so that less locking
+// is needed when iterating over entries to read them.
+type reflogRingBuffer struct {
+	items         []reflogRootHashEntry
+	mu            *sync.Mutex
+	requestedSize int
+	totalSize     int
+	insertIndex   int
+	itemCount     int
+	epoch         uint
+}
+
+// newReflogRingBuffer creates a new reflogRingBuffer that allows the reflog to query up to |size| records.
+// Internally, the ring buffer allocates extra storage so that |size| records can be read while new root entries
+// are still being recorded.
+func newReflogRingBuffer(size int) *reflogRingBuffer {
+	if size < 0 {
```

Comment:
<=?

Reply:
I actually do allow 0 to be passed in – if the `DOLT_DISABLE_REFLOG` env var is set, then the reflog ring buffer is still created, but created with a 0 record capacity. It could be changed to not instantiate the reflog ring buffer in that case if we do want to change this.